Walk Python AST to check for unsupported attribute type annotations

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51805 Walk Python AST to check for unsupported attribute type annotations**

Differential Revision: [D26418589](https://our.internmc.facebook.com/intern/diff/D26418589)

Commit:
Walk Python AST to check for unsupported attribute type annotations

[ghstack-poisoned]

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"




[ghstack-poisoned]

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"




[ghstack-poisoned]

suo commented on torch/jit/_check.py:
```diff
@@ -0,0 +1,214 @@
+        # cannot be reassigned later to a non-empty tuple. Same
+        # deal with `NamedTuple`
+
+        containers = {"List", "Dict", "Optional"}
```
Note that matching on names is not necessarily reliable, since `from typing import List as foo`  is possible. But this seems fine, probably no one does that.

suo commented on torch/jit/_check.py:
```diff
@@ -0,0 +1,214 @@
+        """
+        # If this isn't a call to `torch.jit.annotate`
+        try:
+            if (node.func.value.value.id != "torch"
```
is this check reliable? What if I do `from torch import jit` and do `jit.annotate()`? (This [does actually happen](https://fburl.com/codesearch/j8hnbwzv)).

suo replied:
(since this more a best effort thing, maybe it's fine to just match against `torch.jit.annotate` and `jit.annotate`)

suo commented on torch/jit/_check.py:
```diff
@@ -0,0 +1,214 @@
+                    or node.func.attr != "annotate"):
+                try:
+                    self.generic_visit(node)
+                except RuntimeError:
```
Shouldn't we just call `self.generic_visit` and let the error propagate naturally? What is the purpose of the try-except-raise?

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"




[ghstack-poisoned]

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"




[ghstack-poisoned]

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"




[ghstack-poisoned]

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"




[ghstack-poisoned]

suo commented on test/jit/test_scriptmod_ann.py:
```diff
@@ -0,0 +1,251 @@
+                self.x = x
+                return 1
+
+        with warnings.catch_warnings(record=True) as w:
```
there is an `assertWarns` in unittest, we should use that.

suo reviewed (approved):
One minor testing comment but otherwise lgtm! Thanks :)

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"


Differential Revision: [D26418589](https://our.internmc.facebook.com/intern/diff/D26418589)

[ghstack-poisoned]

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"


Differential Revision: [D26418589](https://our.internmc.facebook.com/intern/diff/D26418589)

[ghstack-poisoned]

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"


Differential Revision: [D26418589](https://our.internmc.facebook.com/intern/diff/D26418589)

[ghstack-poisoned]

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"


Differential Revision: [D26418589](https://our.internmc.facebook.com/intern/diff/D26418589)

[ghstack-poisoned]

Commit:
Update on "Walk Python AST to check for unsupported attribute type annotations"


Differential Revision: [D26418589](https://our.internmc.facebook.com/intern/diff/D26418589)

[ghstack-poisoned]

wconstab commented on test/jit/test_scriptmod_ann.py:
```diff
@@ -0,0 +1,251 @@
+
+        with self.assertRaisesRegex(RuntimeError, "Tried to set "
+                                    "nonexistent attribute"):
+            with self.assertWarnsRegex(UserWarning, "doesn't support "
```
just a nit, but you might want to replace these 2 strings with a global
WARN_INSTANCE_ANNOT_EMPTY_NON_BASE_TYPE = ".." and then have the string only defined in one place?

wconstab reviewed (approved):
Nice! Thanks for fixing this. lgtm.
